File: backend/app/schemas/menu.py
"""Menu schemas."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

MenuType = Literal["C", "M", "F"]


class MenuNode(BaseModel):
    """Authorized menu node returned by /auth/menus."""

    id: int
    menuCode: str
    parentId: int | None = None
    type: MenuType
    title: str
    path: str | None = None
    componentPath: str | None = None
    icon: str | None = None
    permissionCode: str | None = None
    order: int = 0
    visible: bool = True
    enabled: bool = True
    tagViewEnabled: bool = True
    keepAliveEnabled: bool = False
    children: list[MenuNode] = Field(default_factory=list)


class SystemMenuItem(MenuNode):
    """Menu node for management endpoints; includes timestamps."""

    createdAt: str | None = None
    updatedAt: str | None = None
    children: list[SystemMenuItem] = Field(default_factory=list)


# Resolve the forward reference so child nodes keep their timestamps.
SystemMenuItem.model_rebuild()

File: frontend/src/types/menu.ts
import type { Id } from './common';

export type MenuType = 'C' | 'M' | 'F';

export interface MenuNode {
  id: Id;
  menuCode: string;
  parentId?: Id | null;
  type: MenuType;
  title: string;
  path?: string | null;
  componentPath?: string | null;
  icon?: string | null;
  permissionCode?: string | null;
  order?: number;
  visible?: boolean;
  enabled?: boolean;
  tagViewEnabled?: boolean;
  keepAliveEnabled?: boolean;
  children: MenuNode[];
}

export interface MenuRouteNode extends MenuNode {
  type: 'M';
  path: string;
  componentPath: string;
}

export interface MenuCatalogNode extends MenuNode {
  type: 'C';
}

export interface MenuFunctionNode extends MenuNode {
  type: 'F';
  permissionCode: string;
}

export interface MenuBreadcrumbItem {
  key: string;
  title: string;
  path?: string;
}

export interface MenuSearchItem {
  key: string;
  title: string;
  path: string;
  labels: string[];
}

export interface MenuOpenKeyMeta {
  key: string;
  parentKey: string | null;
  ancestorKeys: string[];
}

export interface RouteRenderRecord {
  key: string;
  title: string;
  path: string;
  componentPath: string;
  menu: MenuRouteNode;
}

export interface ButtonPermissionRequest {
  path: string;
  permissionCode: string;
}
